protocol/icmpv6.go

```diff
@@ -0,0 +1,643 @@
+package protocol
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+
+	"antrea.io/libOpenflow/util"
+)
+
+const (
+	ICMPv6_Type_EchoRequest  = 128
+	ICMPv6_Type_EchoReply    = 129
+	ICMPv6_Type_MLD_Query    = 130
+	ICMPv6_Type_MLDv2_Report = 143
+	ICMPv6_Type_MLD_Report   = 131
+	ICMPv6_Type_MLD_Done     = 132
+
+	ICMPv6_ErrType_Destination_Unreachable = 1
+	ICMPv6_ErrType_Packet_Large            = 2
+	ICMPv6_ErrType_Timeout                 = 3
+	ICMPv6_ErrType_Parameter               = 4
+
+	// Code for ICMPv6 Error: Destination Unreachable
+	ICMPv6_ErrCode_NoRoute         = 0
+	ICMPv6_ErrCode_AdminProhibited = 1
+	ICMPv6_ErrCode_BeyondOfSource  = 2
+	ICMPv6_ErrCode_AddrUnreachable = 3
+	ICMPv6_ErrCode_PortUnreachable = 4
+	ICMPv6_ErrCode_SourcePolicy    = 5
+	ICMPv6_ErrCode_Reject          = 6
+
+	// Code for ICMPv6 Error: Timeout
+	ICMPv6_ErrCode_TTL              = 0
+	ICMPv6_ErrCode_Fragment_Timeout = 1
+
+	// Code for ICMPv6 Error: Parameter
+	ICMPv6_ErrCode_Err_Header     = 0
+	ICMPv6_ErrCode_Unknown_Header = 1
+	ICMPv6_ErrCode_Unknown_Option = 2
+)
+
+//       0                   1                   2                   3
+//       0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//      |     Type      |     Code      |          Checksum             |
+//      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//      |                                                               |
+//      +                         Message Body                          +
+//      |                                                               |
+type ICMPv6Header struct {
+	Type     uint8
+	Code     uint8
+	Checksum uint16
+}
+
+func (i *ICMPv6Header) Len() uint16 {
+	return 4
+}
+
+func (i *ICMPv6Header) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(i.Len()))
+	data[0] = i.Type
+	data[1] = i.Code
+	binary.BigEndian.PutUint16(data[2:4], i.Checksum)
+	return
+}
+
+func (i *ICMPv6Header) UnmarshalBinary(data []byte) error {
+	if len(data) < 4 {
+		return errors.New("The []byte is too short to unmarshal a full ICMPv6Header message.")
+	}
+	i.Type = data[0]
+	i.Code = data[1]
+	i.Checksum = binary.BigEndian.Uint16(data[2:4])
+	return nil
+}
+
+type ICMPv6EchoReqRpl struct {
+	ICMPv6Header
+	Identifier uint16
+	SeqNum     uint16
+	Data       util.Message
+}
+
+func (i *ICMPv6EchoReqRpl) Len() (n uint16) {
+	n = 8
+	if i.Data != nil {
+		n += i.Data.Len()
+	}
+	return 8 + i.Data.Len()
```

Comment:
Can we just have a 'return' here. 'n' is already correctly assigned.

Reply:
Updated.